### PR TITLE
improvement(cluster_gce): retry on create node error

### PR DESCRIPTION
**Pull request by KnifeyMoloko**
We want our GCE cluster code to retry creating a node in case of bumping into resource limitations. The goal is to retry up to 3 times, with a 15 minutes backoff between retries. This PR abstracts the previous try / catch / retry code into a separate method in the `cluster_gce` module, and uses the `retrying` decorator for the retry logic. 

Trello task: https://trello.com/c/Xb4Aalnz

Jenkins run with 2 db nodes being created on GCE: https://jenkins.scylladb.com/view/staging/job/scylla-staging/job/maciej/job/mc-longevity/job/mc-byo-longevity-test/178/

## PR pre-checks (self review)

- [x] I followed [KISS principle](https://en.wikipedia.org/wiki/KISS_principle) and [best practices](https://docs.google.com/document/d/1jihgOKb5iGRlD8_HQ92O0JbLk1kASUoZT23i_MXFSKI)
- [x] I didn't leave commented-out/debugging code
- [x] I added the relevant `backport` labels
- [ ] ~~New configuration option are added and documented (in `sdcm/sct_config.py`)~~
- [ ] ~~I have added tests to cover my changes (Infrastructure only - under `unit-test/` folder)~~
- [x] All new and existing unit tests passed (CI)
- [ ] ~~I have updated the Readme/doc folder accordingly (if needed)~~

### CODE REVIEW

**Review by vponomaryov — DISMISSED** [review]

File: sdcm/cluster_gce.py [hunk 1]
```diff
@@ -341,42 +341,59 @@ def _create_instance(self, node_index, dc_idx, spot=False):
...
+                                  name: str,
+                                  dc_idx: int,
+                                  create_node_params: dict[str, Any],
+                                  spot: bool = False) -> GCENode:
+        try:
+            instance = self._gce_services[dc_idx].create_node(**create_node_params)
+            return instance
+        except GoogleBaseError as gbe:
+            if not spot:
+                raise gbe
```
Comment (enaydanov) [inline sdcm/cluster_gce.py:371]: It can be just `raise` to reraise a caught exception.

File: sdcm/cluster_gce.py [hunk 2]
```diff
@@ -341,42 +341,59 @@ def _create_instance(self, node_index, dc_idx, spot=False):
...
+
+            #  attempt to destroy if node did not start due to preemption
+            if "Instance failed to start due to preemption" in str(gbe):
+                try:
+                    self._destroy_node(name=name, dc_idx=dc_idx)
+                except ResourceNotFoundError:
+                    LOGGER.warning("Attempted to destroy node: {name: %s, idx:%s}, but could not find it. "
+                                   "Possibly the node was destroyed already.", name, str(dc_idx))
```
Comment (enaydanov) [inline sdcm/cluster_gce.py:379]: Don't need to do explicit `str()` here: it'll be converted to a string by `%s`.

File: sdcm/cluster_gce.py [hunk 3]
```diff
@@ -341,42 +341,59 @@ def _create_instance(self, node_index, dc_idx, spot=False):
...
+            raise gbe
+
     def _create_instances(self, count, dc_idx=0):
         spot = 'spot' in self.instance_provision
         instances = []
         for node_index in range(self._node_index + 1, self._node_index + count + 1):
             instances.append(self._create_instance(node_index, dc_idx, spot))
         return instances
 
+    def _destroy_node(self, name: str, dc_idx: int) -> bool:
```
Comment (enaydanov) [inline sdcm/cluster_gce.py:392]: Do you think `_destroy_instance` will be better name for this method?

**Review by enaydanov — COMMENTED** [review]
So, the logic changed from:
```
  on_demand -> fail
  spot -> on_demand -> fail
```
To
```
  on_demand -> 15m wait -> on_demand -> 15m wait -> on_demand -> fail
  spot -> 15m wait -> on_demand -> 15m wait -> on_demand -> fail
```
?

Am I right?

**Comment by KnifeyMoloko** [comment]
> So, the logic changed from:
> 
> ```
>   on_demand -> fail
>   spot -> on_demand -> fail
> ```
> 
> To
> 
> ```
>   on_demand -> 15m wait -> on_demand -> 15m wait -> on_demand -> fail
>   spot -> 15m wait -> on_demand -> 15m wait -> on_demand -> fail
> ```
> 
> ?
> 
> Am I right?

Yup. We're retrying 3x with a 15m wait regardless of what error we encounter. The only twist is we attempt to destroy a preempted spot instance and move to `on_demand` if we encountered preemption.

File: sdcm/cluster_gce.py (hunk 3 above)
Comment (KnifeyMoloko) [reply sdcm/cluster_gce.py]: Yes, that's better. Fixed.

File: sdcm/cluster_gce.py (hunk 2 above)
Comment (KnifeyMoloko) [reply sdcm/cluster_gce.py]: Removed the cast to `str`. Fixed.

File: sdcm/cluster_gce.py (hunk 1 above)
Comment (KnifeyMoloko) [reply sdcm/cluster_gce.py]: Fixed

File: sdcm/cluster_gce.py [hunk 4]
```diff
@@ -341,42 +341,59 @@ def _create_instance(self, node_index, dc_idx, spot=False):
...
+              message="Retrying to create a GCE node...",
+              raise_on_exceeded=True)
+    def _create_node_with_retries(self,
+                                  name: str,
+                                  dc_idx: int,
+                                  create_node_params: dict[str, Any],
+                                  spot: bool = False) -> GCENode:
+        try:
+            instance = self._gce_services[dc_idx].create_node(**create_node_params)
+            return instance
```
[the remark this answers is not in the document]
Comment (KnifeyMoloko) [reply sdcm/cluster_gce.py]: Fixed (inlined the return statement)

**Review by vponomaryov — APPROVED** [review]

**Pull request merged**